org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/SideEffectChecker.java

```diff
@@ -0,0 +1,324 @@
+	}
+
+
+	public boolean hasSideEffect() {
+		return fSideEffect;
+	}
+
+	@Override
+	public boolean preVisit2(ASTNode node) {
+		if (hasSideEffect()) {
+			return false;
+		}
+		if (selfModied(node)) {
+			fSideEffect= true;
+			return false;
+		}
+		if (node instanceof MethodInvocation) {
+			MethodInvocation node2= (MethodInvocation) node;
+			final IMethodBinding resolveMethodBinding= node2.resolveMethodBinding();
+			if (MARKEDMETHODSET.contains(getQualifiedName(resolveMethodBinding))) {
+				fSideEffect= true;
+				return false;
+			}
+			MethodDeclaration md= findFunctionDefinition(resolveMethodBinding.getDeclaringClass(), resolveMethodBinding);
+			if (md != null && md.getLength() < THRESHOld) {
+				MethodVisitor mv= new MethodVisitor();
+				md.accept(mv);
+				fSideEffect= mv.hasUpdateNoTemp() == true ? true : fSideEffect;
+			}
+		}
+		return super.preVisit2(node);
+	}
+
+	private String getQualifiedName(IMethodBinding imb) {
+		if (imb == null || imb.getDeclaringClass() == null)
+			return null;
+		return imb.getDeclaringClass().getQualifiedName() + "." + imb.getName(); //$NON-NLS-1$
+	}
+
+
+
+	private static MethodDeclaration findFunctionDefinition(ITypeBinding iTypeBinding, IMethodBinding methodBinding) {
+		if (methodBinding == null ||
+				iTypeBinding == null || !(iTypeBinding.getJavaElement() instanceof IType)) {
+			return null;
+		}
+		IType it= (IType) (iTypeBinding.getJavaElement());
+		try {
+			ITypeHierarchy ith= it.newTypeHierarchy(iTypeBinding.getJavaElement().getJavaProject(), null);
+			IMethod iMethod= (IMethod) methodBinding.getJavaElement();
+			if (iMethod == null) {
+				return null;
+			}
+
+			ArrayList<IType> iTypes= new ArrayList<>();
+			findTypes(it, ith, iTypes);
+			MethodDeclaration res= null;
+			for (IType t : iTypes) {
+				IMethod tmp= JavaModelUtil.findMethod(iMethod.getElementName(),
+						iMethod.getParameterTypes(), false, t);
+				if (tmp != null) {
+					ICompilationUnit icu= tmp.getCompilationUnit();
+					if (icu == null || icu.getSource() == null) {
+						return null;
+					}
+					ASTParser parser= ASTParser.newParser(AST.getJLSLatest());
+					parser.setKind(ASTParser.K_COMPILATION_UNIT);
+					parser.setSource(icu);
+					parser.setResolveBindings(true);
+					CompilationUnit compilationUnit= (CompilationUnit) parser.createAST(null);
+					final ASTNode perform= NodeFinder.perform(compilationUnit, tmp.getSourceRange());
+					if (perform instanceof MethodDeclaration) {
+						MethodDeclaration md= (MethodDeclaration) perform;
+						if (Modifier.isAbstract(md.resolveBinding().getModifiers()))
+							continue;
+						if (res != null)
+							return null;
+						res= md;
+					}
+				}
+			}
+			return res;
+		} catch (JavaModelException e) {
+			e.printStackTrace();
+		}
+		return null;
+	}
+
+	private static void findTypes(IType it, ITypeHierarchy ith, ArrayList<IType> iTypes) throws JavaModelException {
+		iTypes.add(it);
+		for (IType i : ith.getAllSubtypes(it)) {
```

Comment:
This can be quite expensive if someone is using a basic Java method.  Have you tried something like a call to an Object method like hashCode()?  You might need to put this in a thread to be able to have it cancellable.  Another possibility is to screen out Java library classes such as java.lang, java.util, java.io, etc... and don't do a type hierarchy in such a case.

Reply:
With the check for JRE prior, this is fine now.